File: CS/Hamburger/HamburgerTemplateSelector.cs
using Hamburger.ViewModel;
using System.Windows;
using System.Windows.Controls;

namespace Hamburger.TemplateSelectors
{
    public class HamburgerMenuItemTemplateSelector : DataTemplateSelector
    {
        public DataTemplate ItemTemplate { get; set; }
        public DataTemplate SubItemTemplate { get; set; }
        public override DataTemplate SelectTemplate(object item, DependencyObject container)
        {
            if (item is HamburgerSubMenuItemViewModel)
                return SubItemTemplate;
            if (item is HamburgerMenuItemViewModel)
                return ItemTemplate;
            return base.SelectTemplate(item, container);
        }
    }
}

File: VB/Hamburger/HamburgerTemplateSelector.vb
Imports Hamburger.ViewModel
Imports System.Windows
Imports System.Windows.Controls

Namespace Hamburger.TemplateSelectors

    Public Class HamburgerMenuItemTemplateSelector
        Inherits DataTemplateSelector

        Public Property ItemTemplate As DataTemplate

        Public Property SubItemTemplate As DataTemplate

        Public Overrides Function SelectTemplate(ByVal item As Object, ByVal container As DependencyObject) As DataTemplate
            If TypeOf item Is HamburgerSubMenuItemViewModel Then Return SubItemTemplate
            If TypeOf item Is HamburgerMenuItemViewModel Then Return ItemTemplate
            Return MyBase.SelectTemplate(item, container)
        End Function
    End Class
End Namespace
